### PR TITLE
IrregSpacedGrid, PiecewiseLinSpacedGrid, PiecewiseLogSpacedGrid, rename existing grids

### LINKED ISSUE #210: Proposed: `ContinuousGrid` hierarchy with `IrregSpacedGrid`
## Motivation

As we add shock discretization functionality (#206), we need to support grids with irregularly spaced points (e.g., Gauss-Hermite quadrature nodes, or grids where we need particular points at a break, say eligibility for a programme and we need points $..., x-\varepsilon, x, ...$). The current `ContinuousGrid` base class requires `start`, `stop`, and `n_points`, which doesn't accommodate irregular spacing.

### 1. Grid Hierarchy

```
ContinuousGrid (base)
├── LinSpacedGrid   (renamed from LinspaceGrid)
├── LogSpacedGrid   (renamed from LogspaceGrid)
└── IrregSpacedGrid (new)
```

### 2. Common Interface

All subclasses would have:
- `values: Float1D` — a `@property` returning the grid points as a JAX array
- `n_points: int` — number of grid points

Additionally:
- `LinSpacedGrid` and `LogSpacedGrid` retain `start` and `stop` attributes
- `IrregSpacedGrid` stores points directly

### 3. `values` vs `to_jax()`

Replace `to_jax()` with a `values` property:
- For `IrregSpacedGrid`, values must be stored at construction time
- For `LinSpacedGrid` / `LogSpacedGrid`, values can be computed lazily on access

This asymmetry affects memory usage and serialization, but a `@property` abstracts it away cleanly.

### 4. Naming: `LinSpacedGrid` over `LinspaceGrid`

The new name reads as "linearly spaced grid" (adjective form) rather than referencing the NumPy function name `linspace`. This is grammatically cleaner and more self-documenting.

## Sketch: `IrregSpacedGrid` Interface

```python
@dataclass(frozen=True, kw_only=True)
class IrregSpacedGrid(ContinuousGrid):
    """A grid of continuous values at irregular (user-specified) points.

    Example:
        Gauss-Hermite quadrature nodes: [-1.73, -0.58, 0.58, 1.73]

    Attributes:
        points: The grid points. Must be a 1D sequence of floats in ascending order.

    """

    points: tuple[float, ...]

    def __post_init__(self) -> None:
        _validate_irreg_spaced_grid(self.points)

    @property
    def n_points(self) -> int:
        return len(self.points)

    @property
    def values(self) -> Float1D:
        return jnp.array(self.points)

    def get_coordinate(self, value: ScalarFloat) -> ScalarFloat:
        # Interpolate position among irregular points
        return grid_helpers.get_irreg_coordinate(value, self.points)
```

## Migration Path

1. Add `values` property to existing grids (can coexist with `to_jax()` initially)
2. Add `IrregSpacedGrid`
3. Rename `LinspaceGrid` → `LinSpacedGrid`, `LogspaceGrid` → `LogSpacedGrid`
4. Deprecate `to_jax()` in favor of `values`

## Related

- #206 — Shock discretization (primary use case for `IrregSpacedGrid`)
- #208 — Categorical decorator / 1D space descriptions

### PR DESCRIPTION
**Pull request by hmgaudecker**
Fix #210, see discussion there.

### CODE REVIEW

**Comment by hmgaudecker** [comment]
@timmens @mj023: Too early for a serious review, but it would be great if you could have a look of whether this is going in the right direction!

File: src/lcm/grids.py [hunk 1]
```diff
@@ -200,6 +203,141 @@ def get_coordinate(self, value: ScalarFloat) -> ScalarFloat:
...
+        """Convert the grid to a Jax array."""
+        piece_arrays = [
+            _generate_piece_points(interval, piece.n_points)
+            for piece, interval in zip(self.pieces, self._parsed_pieces, strict=True)
+        ]
+        return jnp.concatenate(piece_arrays)
+
+    def get_coordinate(self, value: ScalarFloat) -> ScalarFloat:
+        """Get the generalized coordinate of a value in the grid."""
+        return grid_helpers.get_irreg_coordinate(value, self.to_jax())
```
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply src/lcm/grids.py]: Haha, nice crossing of messages! Looks like I wasn't far off 😃

**Comment by mj023** [comment]
A quick heads up, just in case you haven't seen it already, that the `IrregSpacedGrids` are currently not working when used in an actual model. The class probably needs to be a subclass of `ContinuousGrid` and then the attributes need to be shuffled around a bit. Otherwise it will not get picked up in the state space.

Surprisingly, replacing the `LinspacedGrids` in the MY-model with the new irregular ones has absolutely no impact on the runtime. It might be that this is a quirk of the model, but it could also be that O(log n) is actually not that much. For 500 points its only 6 comparisons.

File: tests/conftest.py
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply tests/conftest.py]: Better?

**Comment by mj023** [comment]
> 
> Interesting! Is that for the run including the compilation or for the second/third time you run it?

This was including the compilation and on all subsequent runs. I suppose there might be some free compute time on the GPU, that gets filled by these extra computations. We had a similar case, where precomputing a grid of values and then accessing these from the user functions, took the same amount of time, as directly computing them on the fly.
